transaction/std/src/memo_builder/gift_code_funding_memo_builder.rs

```diff
@@ -0,0 +1,444 @@
+
+    #[test]
+    fn test_gift_code_funding_memo_built_successfully_with_note() {
+        // Create memo builder
+        let mut builder = GiftCodeFundingMemoBuilder::default();
+
+        // Set the gift code TxOut public key and note
+        let mut rng: StdRng = SeedableRng::from_seed([0u8; 32]);
+        let gift_code_public_key = RistrettoPublic::from_random(&mut rng);
+        let note = "It's MEMO TIME!!";
+        builder
+            .set_gift_code_tx_out_public_key(&gift_code_public_key)
+            .unwrap();
+        builder.set_funding_note(note).unwrap();
+
+        // Build the memo payload and get the data
+        let memo_payload = build_zero_value_change_memo(&mut builder).unwrap();
+        let memo_data = memo_payload.get_memo_data();
+
+        // Verify memo data
+        let derived_note = MemoDecoder::decode_funding_note(memo_data);
+        let expected_hash = MemoDecoder::tx_out_public_key_short_hash(&gift_code_public_key);
+        assert_eq!(note, derived_note);
+        assert_eq!(
+            expected_hash,
+            memo_data[0..GiftCodeFundingMemo::HASH_DATA_LEN]
+        );
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_built_successfully_without_note() {
+        // Create memo builder
+        let mut builder = GiftCodeFundingMemoBuilder::default();
+
+        // Set the gift code TxOut public key only
+        let mut rng: StdRng = SeedableRng::from_seed([0u8; 32]);
+        let gift_code_public_key = RistrettoPublic::from_random(&mut rng);
+        builder
+            .set_gift_code_tx_out_public_key(&gift_code_public_key)
+            .unwrap();
+
+        // Build the memo payload and get the data
+        let memo_payload = build_zero_value_change_memo(&mut builder).unwrap();
+        let memo_data = memo_payload.get_memo_data();
+
+        // Verify memo data
+        let derived_note = MemoDecoder::decode_funding_note(memo_data);
+        let expected_hash = MemoDecoder::tx_out_public_key_short_hash(&gift_code_public_key);
+        let blank_note = "";
+        assert_eq!(blank_note, derived_note);
+        assert_eq!(
+            expected_hash,
+            memo_data[0..GiftCodeFundingMemo::HASH_DATA_LEN]
+        );
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_fails_to_build_if_key_matches_empty_change_memo_public_key() {
+        // Create memo builder
+        let mut builder = GiftCodeFundingMemoBuilder::default();
+
+        // Set the gift code note and the public_key as the empty gift code public_key
+        let mut rng: StdRng = SeedableRng::from_seed([0u8; 32]);
+        let change_tx_public_key = RistrettoPublic::from_random(&mut rng);
+        let note = "It's MEMO TIME!!";
+        builder
+            .set_gift_code_tx_out_public_key(&change_tx_public_key)
+            .unwrap();
+        builder.set_funding_note(note).unwrap();
+
+        // Build a memo
+        let mut rng: StdRng = SeedableRng::from_seed([0u8; 32]);
```

Comment:
Is there a reason not to use `build_change_memo_with_amount()`?

Reply:
Yes because in that function, there's no way to specify which public key goes into it. This tests ensures that the public key set onto the memo isn't that of the change output and I needed a way to do that here. Thus why it's the only test that doesn't use build_change_memo_with_amount()